File: 0427-construct-quad-tree/0427-construct-quad-tree.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
}
*/

class Solution {
    private boolean isLeaf(int[][] grid, int x1, int y1, int length) {
        int baseNum = grid[x1][y1];
        for (int i=0; i < length; i++) {
            for (int j=0; j < length; j++) {
                if(grid[i + x1][j + y1] !=  baseNum) {
                    return false;
                }
            }   
        }
        return true;
    }
    private Node solve(int[][] grid, int x1, int y1, int length) {
        if(isLeaf(grid, x1, y1, length)) {
            System.out.println(grid[x1][y1]);
            return new Node(grid[x1][y1] == 1, true);
        } else {
            Node node = new Node(false, false);
    
        node.topLeft = solve(grid, x1, y1, length/2);
        node.topRight = solve(grid, x1 , y1 + length / 2, length/2);
        node. bottomLeft = solve(grid, x1 + length / 2, y1, length/2);
        node. bottomRight = solve(grid, x1 + length / 2, y1 + length / 2, length/2);

            return node;
        }
    }

    public Node construct(int[][] grid) {
        return solve(grid, 0, 0, grid.length);
    }
}

File: 0427-construct-quad-tree/0427-construct-quad-tree.py
"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution:
    def is_leaf(self, grid: List[List[int]], i: int, j: int, size:int) -> bool:
        curr = grid[i][j]
        for x in range(i, i + size): 
            for y in range(j, j+ size): 
                if grid[x][y] != curr: 
                    return False
        
        return True

    def construct(self, grid: List[List[int]]) -> 'Node':   
        i =  j = 0
        size = len(grid)
        node = self.build(grid, i, j, size)
        return node

    def build(self, grid: List[List[int]], i: int, j: int, size: int) -> 'Node':
        is_leaf = self.is_leaf(grid, i, j, size)
        if is_leaf: 
            node = Node(grid[i][j], True)
            return node
        
        node = Node(False, False)
        half = size // 2
        node.topLeft = self.build(grid, i, j , half)
        node.topRight = self.build(grid, i, j  + half, half)
        node.bottomLeft = self.build(grid, i + half, j , half)
        node.bottomRight = self.build(grid, i + half, j + half , half)
        return node
        

